cli/cmd/endpoints.go

```diff
@@ -86,17 +97,88 @@ requests.`,
 		},
 	}
 
-	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified endpoints (default: all namespaces)")
 	cmd.PersistentFlags().StringVarP(&options.outputFormat, "output", "o", options.outputFormat, fmt.Sprintf("Output format; one of: \"%s\" or \"%s\"", tableOutput, jsonOutput))
 
 	return cmd
 }
 
-func requestEndpointsFromAPI(client public.APIClient) (*pb.EndpointsResponse, error) {
-	return client.Endpoints(context.Background(), &pb.EndpointsParams{})
+func requestEndpointsFromAPI(client public.APIClient, authorities []string) (endpointsInfo, error) {
+	info := make(endpointsInfo)
+	// buffered channels to avoid blocking
+	events := make(chan *destinationPb.Update, len(authorities))
+	errors := make(chan error, len(authorities))
+	var wg sync.WaitGroup
+
+	for _, authority := range authorities {
+		wg.Add(1)
+		go func(authority string) {
+			defer wg.Done()
+			if len(errors) == 0 {
+				dest := &destinationPb.GetDestination{
+					Scheme: "http:",
+					Path:   authority,
+				}
+
+				rsp, err := client.Get(context.Background(), dest)
+				if err != nil {
+					errors <- err
+					return
+				}
+
+				event, err := rsp.Recv()
+				if err != nil {
+					errors <- err
+					return
+				}
+				events <- event
+			}
+		}(authority)
+	}
+	// Block till all goroutines above are done
+	wg.Wait()
+
+	for i := 0; i < len(authorities); i++ {
+		select {
+		case err := <-errors:
+			// we only care about the first error
```

Comment:
I think the blocking on `wg.Wait()` is not a big deal because the loop on authorities is really quick (it just spawns goroutines).
In your change, that loop is gonna run pretty quickly as well, and then it'll close the events channel. The nested go routines will then probably panic when they try to sending to it.
We could indeed use a third channel to exit on the first failure and avoid extra api calls after that, but I wouldn't use a new context for that purpose. If we wanted to set a context timeout we should pass that into the `client.Get` call and I think the timeout would be caught by that call returning an error.

Reply:
I'll probably just add a `if ctx.Err() != nil` check before sending any more messages to the events channel. But yeah, it's a TIOLI. What you have here is good :+1:.